cmd/oras/internal/option/format.go

```diff
@@ -15,21 +15,120 @@ limitations under the License.
 
 package option
 
-import "github.com/spf13/pflag"
+import (
+	"bytes"
+	"fmt"
+	"strings"
+	"text/tabwriter"
 
-// Format is a flag to format metadata into output.
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+	oerrors "oras.land/oras/cmd/oras/internal/errors"
+)
+
+// FormatType represents a custom description in help doc.
+type FormatType struct {
+	Name  string
+	Usage string
+}
+
+// WithUsage returns a new format type with provided usage string.
+func (ft *FormatType) WithUsage(usage string) *FormatType {
+	return &FormatType{
+		Name:  ft.Name,
+		Usage: usage,
+	}
+}
+
+// format types
+var (
+	FormatTypeJSON = &FormatType{
+		Name:  "json",
+		Usage: "Print in JSON format",
+	}
+	FormatTypeGoTemplate = &FormatType{
+		Name:  "go-template",
+		Usage: "Print in JSON format",
+	}
+	FormatTypeTable = &FormatType{
+		Name:  "table",
+		Usage: "Get direct referrers and output in table format",
+	}
+	FormatTypeTree = &FormatType{
+		Name:  "tree",
+		Usage: "Get referrers recursively and print in tree format",
+	}
+)
+
+// Format contains input and parsed options for formatted output flags.
 type Format struct {
-	Template string
+	FormatFlag   string
+	Type         string
+	Template     string
+	AllowedTypes []*FormatType
 }
 
 // ApplyFlag implements FlagProvider.ApplyFlag.
 func (opts *Format) ApplyFlags(fs *pflag.FlagSet) {
-	const name = "format"
-	if fs.Lookup(name) != nil {
-		// allow command to overwrite the flag
-		return
-	}
-	fs.StringVar(&opts.Template, name, "", `[Experimental] Format output using a custom template:
-'json':       Print in JSON format
-'$TEMPLATE':  Print output using the given Go template.`)
+	var buf bytes.Buffer
+	_, _ = buf.WriteString("[Experimental] Format output using a custom template:")
+	w := tabwriter.NewWriter(&buf, 0, 0, 2, ' ', 0)
+	for _, t := range opts.AllowedTypes {
+		_, _ = fmt.Fprintf(w, "\n'%s':\t%s", t.Name, t.Usage)
+	}
+	w.Flush()
+	// apply flags
+	fs.StringVar(&opts.FormatFlag, "format", opts.FormatFlag, buf.String())
+	fs.StringVar(&opts.Template, "template", "", "[Experimental] Template string used to format output")
+}
+
+// Parse parses the input format flag.
+func (opts *Format) Parse(_ *cobra.Command) error {
+	if err := opts.parseFlag(); err != nil {
+		return err
+	}
+
+	if opts.Type == "" {
+		// flag not specified
+		return nil
+	}
+
+	if opts.Type == FormatTypeGoTemplate.Name && opts.Template == "" {
+		return &oerrors.Error{
+			Err:            fmt.Errorf("%q format specified but no template given", opts.Type),
+			Recommendation: fmt.Sprintf("use `--format %s=TEMPLATE` or `--format %s --template TEMPLATE` to specify the template", opts.Type, opts.Type),
```

Comment:
We should only recommend the primary usage
```suggestion
			Recommendation: fmt.Sprintf("use `--format %s=TEMPLATE` to specify the template", opts.Type, opts.Type),
```